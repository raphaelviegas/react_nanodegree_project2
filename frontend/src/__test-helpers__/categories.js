export const initialState = { categories: [] }

export const categories = [
    {
        name: 'react',
        path: 'react'
    },
    {
        name: 'redux',
        path: 'redux'
    },
    {
        name: 'udacity',
        path: 'udacity'
    }
]