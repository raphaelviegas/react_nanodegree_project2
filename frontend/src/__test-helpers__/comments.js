const uuidv4 = require('uuid/v4')

export const initialState = { comments: [] }

export const comments = [
    {
        id: '894tuq4ut84ut8v4t8wun89g',
        parentId: "8xf0y6ziyjabvozdd253nd",
        timestamp: 1468166872634,
        body: 'Hi there! I am a COMMENT.',
        author: 'thingtwo',
        voteScore: 6,
        deleted: false,
        parentDeleted: false
    },
    {
        id: '8tu4bsun805n8un48ve89',
        parentId: "8xf0y6ziyjabvozdd253nd",
        timestamp: 1469479767190,
        body: 'Comments. Are. Cool.',
        author: 'thingone',
        voteScore: -5,
        deleted: false,
        parentDeleted: false
    }
]

export const getCommentsId = '894tuq4ut84ut8v4t8wun89g'

export const removeComment = '894tuq4ut84ut8v4t8wun89g'

export const voteComment = {
    id: '894tuq4ut84ut8v4t8wun89g',
    voteScore: 10
}

let id = uuidv4()
export const addCommentInfo = {
    parentId: '8xf0y6ziyjabvozdd253nd',
    author: 'Test Author',
    body: 'Test Body',
    id
}

export const editCommentInfo = {
    id: '894tuq4ut84ut8v4t8wun89g',
    editContent: {
        body: 'Edited comment body',
        timestamp: Date.now()
    }
}