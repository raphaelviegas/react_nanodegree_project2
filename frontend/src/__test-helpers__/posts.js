const uuidv4 = require('uuid/v4')

export const initialState = {
    posts: [],
    sortByOption: '-voteScore',
    invalidPost: false,
    loadingPosts: true,
}

export const posts = [
    {
        id: '8xf0y6ziyjabvozdd253nd',
        timestamp: 1467166872634,
        title: 'Udacity is the best place to learn React',
        body: 'Everyone says so after all.',
        author: 'thingtwo',
        category: 'react',
        voteScore: 10,
        deleted: false,
        commentCount: 2
    },
    {
        id: '6ni6ok3ym7mf1p33lnez',
        timestamp: 1468479767190,
        title: 'Learn Redux in 10 minutes!',
        body: 'Just kidding. It takes more than 10 minutes to learn technology.',
        author: 'thingone',
        category: 'redux',
        voteScore: -5,
        deleted: false,
        commentCount: 0
    }
]

export const removePost = '8xf0y6ziyjabvozdd253nd'

export const votePost = {
    id: '8xf0y6ziyjabvozdd253nd',
    voteScore: 10
}

let id = uuidv4()
export const addPostInfo = {
    author: "Test Author",
    body: "Test Body",
    category: "Test",
    title: "Test Title",
    timestamp: Date.now(),
    id
}

export const editPostInfo = {
    id: '8xf0y6ziyjabvozdd253nd',
    editContent: {
        title: 'Edited post title',
        body: 'Edited post body'
    }
}