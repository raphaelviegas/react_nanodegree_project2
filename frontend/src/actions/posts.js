import * as ServerAPI from '../helpers/ServerAPI'

export const IS_LOADING = 'IS_LOADING'
export const LOAD_POSTS = 'LOAD_POSTS'
export const REMOVE_POST = 'REMOVE_POST'
export const VOTE_POST = 'VOTE_POST'
export const ORDER_POSTS = 'ORDER_POSTS'
export const ADD_POST = 'ADD_POST'
export const EDIT_POST = 'EDIT_POST'

//Local state
export function isLoading() {
  return {
    type: IS_LOADING
  }
}

export function loadPosts(posts) {
  return {
    type: LOAD_POSTS,
    payload: posts
  }
}

export function votePost(id, voteScore) {
  return {
    type: VOTE_POST,
    payload: { id, voteScore }
  }
}

export function removePost(id) {
  return {
    type: REMOVE_POST,
    payload: id
  }
}

export function orderPosts(orderOption) {
  return {
    type: ORDER_POSTS,
    payload: orderOption
  }
}

export function addPost(post) {
  return {
    type: ADD_POST,
    payload: post
  }
}

export function editPost(id, editContent) {
  return {
    type: EDIT_POST,
    payload: {
      id,
      editContent
    }
  }
}

//Fetch requests
export const fetchPosts = () => dispatch => (
  ServerAPI.getPosts()
    .then(posts => dispatch(loadPosts(posts)))
)

export const fetchVotePost = (id, option) => dispatch => (
  ServerAPI.vote('posts', id, option)
    .then(post => dispatch(votePost(post.id, post.voteScore)))
)

export const fetchRemovePost = (id) => dispatch => (
  ServerAPI.remove('posts', id)
    .then(post => dispatch(removePost(post.id)))
)

export const fetchAddPost = (postInfo) => dispatch => (
  ServerAPI.add('posts', postInfo)
    .then(post => dispatch(addPost(post)))
)

export const fetchEditPost = (id, editContent) => dispatch => (
  ServerAPI.edit('posts', id, editContent)
    .then(() => dispatch(editPost(id, editContent)))
)





