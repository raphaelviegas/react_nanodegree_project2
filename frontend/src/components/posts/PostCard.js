import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import MaterialIcon from 'material-icons-react'
import VoteScore from '../shared/VoteScore'
import EditPostForm from '../posts/editPost/EditPostForm'

class PostCard extends Component {
    static propTypes = {
        post: PropTypes.object.isRequired,
        onRemovePost: PropTypes.func.isRequired
    }

    state = {
        showEdit: '',
    }

    showEdit = () => this.setState(() => ({ showEdit: 'card-reveal-active' }))
    hideEdit = () => this.setState(() => ({ showEdit: '' }))

    removePost(id) {
        this.props.onRemovePost(id)
    }
    render() {
        const { id, title, body, author, voteScore, commentCount, category } = this.props.post
        const { showEdit } = this.state
        return (
            <div className="col s12 m6">
                <div className="card grey lighten-4">
                    <div className="card-content">
                        <span className="card-title truncate">{title}</span>
                        <span className="card-author">{author}</span>
                        <p>{body}</p>
                    </div>
                    <div className="card-action valign-wrapper">
                        <VoteScore size={16} score={voteScore} type='post' id={id} />
                        <Link className="align-right" to={`/${category}/${id}`}>Show Post</Link>
                    </div>
                    <div className="card-comments">
                        <span className="card-comments-number">{commentCount}</span>
                    </div>
                    <div className="card-buttons">
                        <div className="card-button-edit" onClick={() => this.showEdit()}>
                            <MaterialIcon icon="edit" size={14} />
                        </div>
                        <div className="card-button-delete" onClick={() => this.removePost(id)}>
                            <MaterialIcon icon="delete" size={14} />
                        </div>
                    </div>
                    <div className={`card-reveal ${showEdit}`}>
                        <EditPostForm
                            id={id}
                            defaultTitle={title}
                            defaultBody={body}
                            onPostCardFormSubmited={this.hideEdit}
                        />
                        <div className="hide-edit" onClick={() => this.hideEdit()}>
                            <MaterialIcon icon="close" color="#c3c3c3" />
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

export default PostCard
