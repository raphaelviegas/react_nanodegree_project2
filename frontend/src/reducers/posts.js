import {
    IS_LOADING,
    LOAD_POSTS,
    REMOVE_POST,
    VOTE_POST,
    ORDER_POSTS,
    ADD_POST,
    EDIT_POST
} from '../actions/posts'

function postReducer(state = {
    posts: [],
    sortByOption: '-voteScore',
    invalidPost: false,
    loadingPosts: true,
}, action) {
    switch (action.type) {
        case IS_LOADING: {
            state = {
                ...state,
                loadingPosts: true
            }
            return state
        }
        case LOAD_POSTS:
            state = {
                ...state,
                loadingPosts: false,
                posts: action.payload
            }
            return state
        case REMOVE_POST:
            state = {
                ...state,
                posts: state.posts.filter(post => post.id !== action.payload)
            }
            return state
        case VOTE_POST:
            state = {
                ...state,
                posts: state.posts.map(post => {
                    if (post.id === action.payload.id) {
                        return {
                            ...post,
                            voteScore: action.payload.voteScore
                        }
                    }
                    return post
                }),
            }
            return state
        case ORDER_POSTS:
            state = {
                ...state,
                sortByOption: action.payload
            }
            return state
        case ADD_POST:
            let posts = state.posts.concat([action.payload])
            state = {
                ...state,
                loadingPosts: false,
                posts
            }
            return state
        case EDIT_POST:
            state = {
                ...state,
                posts: state.posts.map(post => {
                    if (post.id === action.payload.id) {
                        post.title = action.payload.editContent.title
                        post.body = action.payload.editContent.body
                    }
                    return post
                })
            }
            return state
        default:
            return state
    }
}

export default postReducer