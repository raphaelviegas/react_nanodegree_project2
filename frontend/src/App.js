import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { fetchPosts } from './actions/posts'
import TopBar from './components/main/TopBar'
import Menu from './components/main/Menu'
import AddPostButton from './components/main/AddPostButton'
import Dynamic from './components/main/Dynamic'

export class App extends Component {

  componentDidMount() {
    this.props.fetchPosts()
  }

  render() {
    return (
      <div className="App">
        <TopBar />
        <div className="row main-row">
          <div className="col s12 m3 leftside-menu">
            <Menu />
          </div>
          <div className="col s12 m9 right-side-content">
            <Dynamic loading={this.props.loadingPosts} />
          </div>
        </div>
        <AddPostButton />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    loadingPosts: state.postsReducer.loadingPosts
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    fetchPosts: () => dispatch(fetchPosts())
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(App));
