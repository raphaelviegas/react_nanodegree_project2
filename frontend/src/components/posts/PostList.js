import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { fetchRemovePost } from '../../actions/posts'
import PostCard from './PostCard'
import sortBy from 'sort-by'

export class PostList extends Component {
  static propTypes = {
    posts: PropTypes.array.isRequired,
    sortByOption: PropTypes.string.isRequired
  }

  removePost = (id) => {
    this.props.fetchRemovePost(id)
  }

  render() {
    const { posts, sortByOption } = this.props
    //SortBy
    const sortedPosts = [...posts].sort(sortBy(sortByOption))
    return (
      <div>
        {posts.length === 0
          ? <div className="noPosts"><h4>No posts to show</h4></div>
          : <div className="row">
            {
              sortedPosts.map((post, index) => {
                return <PostCard post={post} onRemovePost={this.removePost} key={index} />
              })
            }
          </div>
        }
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    posts: ownProps.posts,
    sortByOption: state.postsReducer.sortByOption
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    fetchRemovePost: (id) => dispatch(fetchRemovePost(id))
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(PostList));
