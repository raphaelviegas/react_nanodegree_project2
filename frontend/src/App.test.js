import React from 'react'
import { shallow } from 'enzyme'
import { App } from './App'
import { initialState } from './__test-helpers__/posts'

describe('<App />', () => {
    const setup = {
        loadingPosts: initialState.loadingPosts,
        fetchPosts: jest.fn()
    }
    let wrapper
    beforeEach(() => {
        setup.fetchPosts.mockClear()
        wrapper = shallow(<App {...setup} />)
    })

    it('Should render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('Should call fetchPosts when component mounts', () => {
        expect(setup.fetchPosts).toHaveBeenCalledTimes(1)
    })
})