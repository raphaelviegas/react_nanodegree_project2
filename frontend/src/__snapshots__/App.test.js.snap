// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    fetchPosts={[Function]}
    loadingPosts={true}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown />,
        <div
          className="row main-row"
>
          <div
                    className="col s12 m3 leftside-menu"
          >
                    <withRouter(Connect(Menu)) />
          </div>
          <div
                    className="col s12 m9 right-side-content"
          >
                    <Dynamic
                              loading={true}
                    />
          </div>
</div>,
        <Unknown />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col s12 m3 leftside-menu"
>
              <withRouter(Connect(Menu)) />
</div>,
            <div
              className="col s12 m9 right-side-content"
>
              <Dynamic
                            loading={true}
              />
</div>,
          ],
          "className": "row main-row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <withRouter(Connect(Menu)) />,
              "className": "col s12 m3 leftside-menu",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Dynamic
                loading={true}
/>,
              "className": "col s12 m9 right-side-content",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "loading": true,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown />,
          <div
            className="row main-row"
>
            <div
                        className="col s12 m3 leftside-menu"
            >
                        <withRouter(Connect(Menu)) />
            </div>
            <div
                        className="col s12 m9 right-side-content"
            >
                        <Dynamic
                                    loading={true}
                        />
            </div>
</div>,
          <Unknown />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col s12 m3 leftside-menu"
>
                <withRouter(Connect(Menu)) />
</div>,
              <div
                className="col s12 m9 right-side-content"
>
                <Dynamic
                                loading={true}
                />
</div>,
            ],
            "className": "row main-row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <withRouter(Connect(Menu)) />,
                "className": "col s12 m3 leftside-menu",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Dynamic
                  loading={true}
/>,
                "className": "col s12 m9 right-side-content",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "loading": true,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
